### ISSUE TITLE
Performance degradation under concurrency

### ISSUE THREAD

**Issue by dvarrazzo** [psycopg/psycopg#448]
@dvarrazzo  Something bad is happening with concurrency. Increasing the number of concurrent connections makes psycopg performance worse (!). I am talking about both psycopg and psycopg_async.

Here is the modified [script](https://gist.github.com/pwtail/5f056752fed2760bc78d3fea64b6f854) I used.
In the case of sync driver adding concurrency means multiple threads, in the async case - multiple coroutines.

Both psycopg2 and asyncpg behave as expected so using concurrency roughly doubles their performance

_Originally posted by @pwtail in https://github.com/psycopg/psycopg/discussions/411#discussioncomment-4288249_

**Comment by dvarrazzo** [comment]
I have added a `--concurrency` option to the `bench-411.py` script and can verify that concurrent operation seem indeed serialized.

**Comment by dvarrazzo** [comment]
Update: I have been profiling psycopg under concurrency and compare it whith psycopg2. Dumping from my diary:

```
    strace -fTr python tests/scripts/bench-411.py -c1 -n 10000 --no-drop --no-create psycopg > strace-c1 2>&1
    strace -fTr python tests/scripts/bench-411.py -c4 -n 10000 --no-drop --no-create psycopg > strace-c4 2>&1
```

Looking the non-concurrent strace, it is a loop of:

```
     0.000030 sendto(3, "B\0\0\0\34\0_pg3_0\0\0\1\0\1\0\1\0\0\0\2#o\0\1\0\0D\0\0"..., 51, MSG_NOSIGNAL, NULL, 0) = 51 <0.000006>
     0.000018 poll([{fd=3, events=POLLIN|POLLERR}], 1, 100) = 1 ([{fd=3, revents=POLLIN}]) <0.000004>
     0.000014 recvfrom(3, "2\0\0\0\4T\0\0\0\264\0\10id\0\0\2=\273\0\1\0\0\0\27\0\4\377\377\377\377\0"..., 16384, 0, NULL, NULL) = 282 <0.000003>
```

In psycopg2, most often, the interlocked loops don't bother the kernel to grab
the GIL:

```
    [pid T1]      0.000014 sendto(3, "Q\0\0\0\244\nSELECT customer.id, custom"..., 165, MSG_NOSIGNAL, NULL, 0) = 165 <0.000015>
    [pid T1]      0.000028 poll([{fd=3, events=POLLIN|POLLERR}], 1, -1 <unfinished ...>
    [pid T2]      0.000005 sendto(4, "Q\0\0\0\244\nSELECT customer.id, custom"..., 165, MSG_NOSIGNAL, NULL, 0) = 165 <0.000014>
    [pid T2]      0.000028 poll([{fd=4, events=POLLIN|POLLERR}], 1, -1 <unfinished ...>
    [pid T1]      0.000028 <... poll resumed>) = 1 ([{fd=3, revents=POLLIN}]) <0.000057>
    [pid T1]      0.000009 recvfrom(3, "T\0\0\0\264\0\10id\0\0\2=\273\0\1\0\0\0\27\0\4\377\377\377\377\0\0name"..., 16384, 0, NULL, NULL) = 270 <0.000006>
    [pid T2]      0.000020 <... poll resumed>) = 1 ([{fd=4, revents=POLLIN}]) <0.000055>
    [pid T2]      0.000009 recvfrom(4, "T\0\0\0\264\0\10id\0\0\2=\273\0\1\0\0\0\27\0\4\377\377\377\377\0\0name"..., 16384, 0, NULL, NULL) = 270 <0.000007>
```

Looking at a concurrent strace of psycopg 3, with two threads, we have a
sequene of calls that looks like:

```
    [pid T1] s.254246 futex(LOCK-ec, FUTEX_WAKE_PRIVATE, 1) = 1 <0.000005>              # T1 releases the GIL
    [pid T2] s.254256 <... futex resumed>) = 0 <0.000043>                               # T2 acquires the GIL
    [pid T1] s.254261 sendto(3, "B\0\0\0\34\0_pg3_0\0\0\1\0\1\0\1\0\0\0\2\22\34\0\1\0\0D\0\0"..., 51, MSG_NOSIGNAL, NULL, 0 <unfinished ...>
                                                                                        # T1 sends the query
    [pid T2] s.254267 futex(LOCK-f0, FUTEX_WAKE_PRIVATE, 1) = 0 <0.000004>              # T2 releases the GIL, no-one needs it
    [pid T1] s.254276 <... sendto resumed>) = 51 <0.000010>                             # T1 finishes to send the query
    [pid T2] s.254280 recvfrom(4, "2\0\0\0\4T\0\0\0\264\0\10id\0\0\2=\273\0\1\0\0\0\27\0\4\377\377\377\377\0"..., 16384, 0, NULL, NULL) = 282 <0.000005>
                                                                                        # T2 receives data from the socket
    [pid T1] s.254292 poll([{fd=3, events=POLLIN|POLLERR}], 1, 100) = 1 ([{fd=3, revents=POLLIN}]) <0.000005>
                                                                                        # T1 polls for data, data is ready immediately
    [pid T1] s.254308 futex(LOCK-e8, FUTEX_WAIT_BITSET_PRIVATE, 0, {tv_sec=1991074, tv_nsec=232748770}, FUTEX_BITSET_MATCH_ANY <unfinished ...>
                                                                                        # T1 start to acquire the GIL (in wait, checking for signals)
                                                                                        # (??? why does it wait? maybe T2 has acquired without kernel arbitration)
                                                                                        # MAYBE THIS IS NOT THE GIL - different address
    [pid T2] s.254338 futex(LOCK-e8, FUTEX_WAKE_PRIVATE, 1) = 1 <0.000006>              # T2 releases the GIL (before sending)
    [pid T1] s.254349 <... futex resumed>) = 0 <0.000036>                               # T1 acquires the GIL
    [pid T2] s.254353 sendto(4, "B\0\0\0\34\0_pg3_0\0\0\1\0\1\0\1\0\0\0\2\34\302\0\1\0\0D\0\0"..., 51, MSG_NOSIGNAL, NULL, 0 <unfinished ...>
                                                                                        # T2 sends the query
    [pid T1] s.254359 futex(LOCK-f0, FUTEX_WAKE_PRIVATE, 1) = 0 <0.000004>              # T1 releases the GIL (?? the one above wasn't the gil)
    [pid T2] s.254369 <... sendto resumed>) = 51 <0.000010>                             # T2 finishes to send the query
    [pid T1] s.254373 recvfrom(3,  <unfinished ...>                                     # T1 starts to receive data from the network
    [pid T2] s.254383 poll([{fd=4, events=POLLIN|POLLERR}], 1, 100 <unfinished ...>     # T2 polls for data, data is not ready yet
    [pid T1] s.254387 <... recvfrom resumed>"2\0\0\0\4T\0\0\0\264\0\10id\0\0\2=\273\0\1\0\0\0\27\0\4\377\377\377\377\0"..., 16384, 0, NULL, NULL) = 281 <0.000010>
                                                                                        # T1 receives the data it was looking for
    [pid T2] s.254393 <... poll resumed>) = 1 ([{fd=4, revents=POLLIN}]) <0.000006>     # T2 gets the message that data is ready to read
    [pid T2] s.254401 futex(LOCK-ec, FUTEX_WAIT_BITSET_PRIVATE, 0, {tv_sec=1991074, tv_nsec=232841927}, FUTEX_BITSET_MATCH_ANY <unfinished ...>
                                                                                        # T2 starts to wait on LOCK-EC (GIL or not GIL?)
    [pid T1] s.254439 futex(LOCK-ec, FUTEX_WAKE_PRIVATE, 1) = 1 <0.000005>              # T1 releases the LOCK-EC
    [pid T2] s.254449 <... futex resumed>) = 0 <0.000043>                               # ...and T2 acquires it.
    [pid T1] s.254453 sendto(3, "B\0\0\0\34\0_pg3_0\0\0\1\0\1\0\1\0\0\0\2#\262\0\1\0\0D\0\0"..., 51, MSG_NOSIGNAL, NULL, 0 <unfinished ...>
                                                                                        # T1 sends a query, and things seem to restart.
```

the futex calls mention three addresses: one ending in F0, two are 4 bytes and
8 bites before that. I should look in the GIL structure if I can figure out
what they are. One of them is more used than the other, but all of them are
exercised pretty heavily. Note that they are 3 locks even when there are more
than 2 threads involved.

```
    $ egrep 'pid.*futex\(' strace-c2 | awk '{print $4}' | sort | uniq -c | sort -nr | head -3
      43941 futex(LOCK-f0,
      24668 futex(LOCK-e8,
      24652 futex(LOCK-ec,
```

What about psycopg2? The calls pattern is the same, and there are the same
three addresses mentioned. The two addresses less used are called about 36% of
the most used, whereas in pg 3 they are called about 52% of the most used one.

```
    strace -fTr python tests/scripts/bench-411.py -c2 -n 1000 --no-drop --no-create psycopg 2>&1 >/dev/null | egrep 'pid.*futex\(' | awk '{print $4}' | sort | uniq -c | sort -nr | head -4
       5227 futex(0x55ab66afabf0,
       2635 futex(0x55ab66afabec,
       2634 futex(0x55ab66afabe8,
          4 futex(0x55ab680f0e50,

    strace -fTr python tests/scripts/bench-411.py -c2 -n 1000 --no-drop --no-create psycopg2 2>&1 >/dev/null | egrep 'pid.*futex\(' | awk '{print $4}' | sort | uniq -c | sort -nr | head -4
       2591 futex(0x55b73dfa6bf0,
        899 futex(0x55b73dfa6be8,
        880 futex(0x55b73dfa6bec,
          4 futex(0x55b73eaacb00,
```

Hypothesis: in pg3 there is higher pid contention because, whenever we release
the gil, we also yield. If that's the case, maybe the `no-generators` branch
might solve the problem.

To a first test, performance increases, although not as much as I'd like.
However it could be a way to proceed, and maybe there are optimizations we can
make.

psycopg2:

```
    $ python tests/scripts/bench-411.py -c4 -n 20000 --no-drop --no-create psycopg2
    2022-12-26 03:50:12,317 INFO Running psycopg2
    2022-12-26 03:50:12,317 INFO thread 0 running 20000 queries
    2022-12-26 03:50:12,319 INFO thread 1 running 20000 queries
    2022-12-26 03:50:12,321 INFO thread 2 running 20000 queries
    2022-12-26 03:50:12,323 INFO thread 3 running 20000 queries
    2022-12-26 03:50:13,755 INFO Run psycopg2 in 1.4314240759704262 s
    2022-12-26 03:50:13,788 INFO Run psycopg2 in 1.4617018748540431 s
    2022-12-26 03:50:13,801 INFO Run psycopg2 in 1.4760340799111873 s
    2022-12-26 03:50:13,815 INFO Run psycopg2 in 1.4895963850431144 s
```

psycopg 3 branch fix-448 (e5431c5)

```
    $ python tests/scripts/bench-411.py -c4 -n 20000 --no-drop --no-create psycopg
    2022-12-26 03:48:47,898 INFO Running psycopg sync
    2022-12-26 03:48:47,898 INFO thread 0 running 20000 queries
    2022-12-26 03:48:47,899 INFO thread 1 running 20000 queries
    2022-12-26 03:48:47,902 INFO thread 2 running 20000 queries
    2022-12-26 03:48:47,902 INFO thread 3 running 20000 queries
    2022-12-26 03:48:54,275 INFO Run psycopg in 6.367022149031982 s
    2022-12-26 03:48:54,308 INFO Run psycopg in 6.399825313128531 s
    2022-12-26 03:48:54,415 INFO Run psycopg in 6.51078495499678 s
    2022-12-26 03:48:54,417 INFO Run psycopg in 6.507864675018936 s
```

psycopg 3 branch no-generators on top of the above (adea9ce)

```
    python tests/scripts/bench-411.py -c4 -n 20000 --no-drop --no-create psycopg
    2022-12-26 03:49:58,338 INFO Running psycopg sync
    2022-12-26 03:49:58,338 INFO thread 0 running 20000 queries
    2022-12-26 03:49:58,341 INFO thread 1 running 20000 queries
    2022-12-26 03:49:58,343 INFO thread 2 running 20000 queries
    2022-12-26 03:49:58,345 INFO thread 3 running 20000 queries
    2022-12-26 03:50:02,369 INFO Run psycopg in 4.024285650812089 s
    2022-12-26 03:50:02,417 INFO Run psycopg in 4.069659866159782 s
    2022-12-26 03:50:02,466 INFO Run psycopg in 4.123257867991924 s
    2022-12-26 03:50:02,509 INFO Run psycopg in 4.160278336144984 s
```

The strace analysis of the locks still shows a higher use of the locks, but
less so. Let's see if a strace shows a different pattern.

Some shotgun attempts: dropping all the GIL releases, except the one in wait()
(because we understand that all the calls are non-blocking, right?)

```
    [pid T1]      0.000033 sendto(3, "B\0\0\0\34\0_pg3_0\0\0\1\0\1\0\1\0\0\0\2\0\201\0\1\0\0D\0\0"..., 51, MSG_NOSIGNAL, NULL, 0) = 51 <0.000005>
    [pid T1]      0.000015 futex(LOCKe8, FUTEX_WAKE_PRIVATE, 1) = 1 <0.000003>
    [pid T2]      0.000007 <... futex resumed>) = 0 <0.000057>
    [pid T1]      0.000002 poll([{fd=3, events=POLLIN|POLLERR}], 1, 100 <unfinished ...>
    [pid T2]      0.000006 futex(LOCKf0, FUTEX_WAKE_PRIVATE, 1 <unfinished ...>
    [pid T1]      0.000002 <... poll resumed>) = 1 ([{fd=3, revents=POLLIN}]) <0.000005>
    [pid T2]      0.000003 <... futex resumed>) = 0 <0.000003>
    [pid T1]      0.000002 futex(LOCKec, FUTEX_WAIT_BITSET_PRIVATE, 0, {tv_sec=2043391, tv_nsec=320551403}, FUTEX_BITSET_MATCH_ANY <unfinished ...>
    [pid T2]      0.000005 recvfrom(5, "2\0\0\0\4T\0\0\0\264\0\10id\0\0\2=\273\0\1\0\0\0\27\0\4\377\377\377\377\0"..., 16384, 0, NULL, NULL) = 275 <0.000005>
    [pid T2]      0.000060 sendto(5, "B\0\0\0\34\0_pg3_0\0\0\1\0\1\0\1\0\0\0\2\1r\0\1\0\0D\0\0"..., 51, MSG_NOSIGNAL, NULL, 0) = 51 <0.000011>
    [pid T2]      0.000027 futex(LOCKec, FUTEX_WAKE_PRIVATE, 1) = 1 <0.000005>
    [pid T1]      0.000008 <... futex resumed>) = 0 <0.000099>
    [pid T2]      0.000004 poll([{fd=5, events=POLLIN|POLLERR}], 1, 100 <unfinished ...>
    [pid T1]      0.000003 futex(LOCKf0, FUTEX_WAKE_PRIVATE, 1 <unfinished ...>
    [pid T2]      0.000005 <... poll resumed>) = 1 ([{fd=5, revents=POLLIN}]) <0.000005>
    [pid T1]      0.000002 <... futex resumed>) = 0 <0.000006>
    [pid T2]      0.000004 futex(LOCKe8, FUTEX_WAIT_BITSET_PRIVATE, 0, {tv_sec=2043391, tv_nsec=320673022}, FUTEX_BITSET_MATCH_ANY <unfinished ...>
    [pid T1]      0.000003 recvfrom(3, "2\0\0\0\4T\0\0\0\264\0\10id\0\0\2=\273\0\1\0\0\0\27\0\4\377\377\377\377\0"..., 16384, 0, NULL, NULL) = 272 <0.000003>
```

If we use select instead of epoll:

```
    [pid T1]      0.000038 sendto(3, "B\0\0\0\34\0_pg3_0\0\0\1\0\1\0\1\0\0\0\2\0\361\0\1\0\0D\0\0"..., 51, MSG_NOSIGNAL, NULL, 0) = 51 <0.000014>
    [pid T1]      0.000025 futex(LOCKec, FUTEX_WAKE_PRIVATE, 1) = 1 <0.000003>
    [pid T2]      0.000007 <... futex resumed>) = 0 <0.000071>
    [pid T1]      0.000002 pselect6(4, [3], [], [3], {tv_sec=0, tv_nsec=0}, NULL <unfinished ...>
    [pid T2]      0.000004 futex(LOCKf0, FUTEX_WAKE_PRIVATE, 1 <unfinished ...>
    [pid T1]      0.000002 <... pselect6 resumed>) = 1 (in [3], left {tv_sec=0, tv_nsec=0}) <0.000003>
    [pid T2]      0.000003 <... futex resumed>) = 0 <0.000004>
    [pid T1]      0.000002 futex(LOCKe8, FUTEX_WAIT_BITSET_PRIVATE, 0, {tv_sec=2044593, tv_nsec=517022482}, FUTEX_BITSET_MATCH_ANY <unfinished ...>
    [pid T2]      0.000006 recvfrom(5, "2\0\0\0\4T\0\0\0\264\0\10id\0\0\2=\273\0\1\0\0\0\27\0\4\377\377\377\377\0"..., 16384, 0, NULL, NULL) = 275 <0.000005>
    [pid T2]      0.000059 sendto(5, "B\0\0\0\34\0_pg3_0\0\0\1\0\1\0\1\0\0\0\2\3U\0\1\0\0D\0\0"..., 51, MSG_NOSIGNAL, NULL, 0) = 51 <0.000011>
    [pid T2]      0.000025 futex(LOCKe8, FUTEX_WAKE_PRIVATE, 1) = 1 <0.000005>
    [pid T1]      0.000008 <... futex resumed>) = 0 <0.000097>
    [pid T2]      0.000004 pselect6(6, [5], [], [5], {tv_sec=0, tv_nsec=0}, NULL <unfinished ...>
    [pid T1]      0.000003 futex(LOCKf0, FUTEX_WAKE_PRIVATE, 1 <unfinished ...>
    [pid T2]      0.000005 <... pselect6 resumed>) = 1 (in [5], left {tv_sec=0, tv_nsec=0}) <0.000005>
    [pid T1]      0.000002 <... futex resumed>) = 0 <0.000006>
    [pid T2]      0.000005 futex(LOCKec, FUTEX_WAIT_BITSET_PRIVATE, 0, {tv_sec=2044593, tv_nsec=517142043}, FUTEX_BITSET_MATCH_ANY <unfinished ...>
    [pid T1]      0.000003 recvfrom(3, "2\0\0\0\4T\0\0\0\264\0\10id\0\0\2=\273\0\1\0\0\0\27\0\4\377\377\377\377\0"..., 16384, 0, NULL, NULL) = 272 <0.000003>
```

end of brain dump. Conclusions later.

**Comment by dvarrazzo** [comment]
What I think is that psycopg is penalised by the way that generators are modularised, and by the fact of using generators.

In psycopg2, when we go non-green and use PQexec, the GIL is released for the whole length of:

- executing the query (aka preparing the messages, `PQsendQueryParams` and similar) 
- flushing the query to the server (`PQflush()`)
- waiting for the result (`PQisReady()/PQconsumeInput()/poll()`)
- receiving the result (`PQgetResult()`)

in psycopg 3, the first point (which profiling shows taking time) is separated from the others because the flushing/waiting/receiving cycle is common to all the `PQsend*()` functions. That's nice from the code organization PoV, but results in re-acquiring the GIL between each of those operations.

We also re-acquire the GIL because every time we need to block we `yield` a wait condition. This is already identified as a possible performance hit (see #405); but it seems that the effect of using generators/wait is especially problematic in multithread code.

So my approach is a fairly major reorganization of the codebase (which should have no visible effect to the end user, so I'm considering it for 3.2):

- re-organise generators in higher-level generators, allowing to release the GIL for longer (#469)
- drop the generators approach and use blocking function which might be executed without the GIL (#470).

**Comment by dlax** [comment]
> So my approach is a fairly major reorganization of the codebase [...]

I'd like to review both the problem statement and proposed solutions; but given how significant this is, it might take some time for me to get through this.

**Comment by dvarrazzo** [comment]
@dlax Another pair of eyes on this problem would be very welcome :) If you want to have a catch up sometimes just let me know.

**Comment by dvarrazzo** [comment]
@dlax heads up: I have tested a bit more, and I'm starting to believe that the problem described here only happens in unrealistic conditions: a tight loop of very fast queries (< 0.1 ms) on localhost. As soon as the query takes a bit longer (e.g. adding a ms pause using `pg_sleep(0.001)`), performances with psycopg2, 3, async or not, asyncpg, aiopg... everything is just dominated by the latency, and parallel execution is no more a problem.

```
(.venv) piro@sherekhan:~/dev/psycopg3$ git log --oneline -1
85151aff (HEAD -> master) fix: raise known errors before checking for signals in waiting

(.venv) piro@sherekhan:~/dev/psycopg3$ git diff -- tests/
diff --git a/tests/scripts/bench-411.py b/tests/scripts/bench-411.py
index a4d1af30..f0cba6ef 100644
--- a/tests/scripts/bench-411.py
+++ b/tests/scripts/bench-411.py
@@ -116,7 +116,7 @@ INSERT INTO customer (id, name, description, q, p, x, y) VALUES
 
 select = """
 SELECT customer.id, customer.name, customer.description, customer.q,
-    customer.p, customer.x, customer.y, customer.z
+    customer.p, customer.x, customer.y, customer.z, pg_sleep(0.001)
 FROM customer
 WHERE customer.id = %(id)s
 """

(.venv) piro@sherekhan:~/dev/psycopg3$ python tests/scripts/bench-411.py -c1 -n 100 --no-drop --no-create psycopg2 psycopg asyncpg
2023-01-01 16:21:12,706 INFO Running psycopg2
2023-01-01 16:21:12,706 INFO thread 0 running 100 queries
2023-01-01 16:21:12,985 INFO Run psycopg2 in 0.2766313059255481 s
2023-01-01 16:21:13,052 INFO Running psycopg sync
2023-01-01 16:21:13,052 INFO thread 0 running 100 queries
2023-01-01 16:21:13,320 INFO Run psycopg in 0.2652683537453413 s
2023-01-01 16:21:13,347 INFO Running asyncpg
2023-01-01 16:21:13,347 INFO task 0 running 100 queries
2023-01-01 16:21:13,654 INFO Run asyncpg in 0.2981667360290885 s
(.venv) piro@sherekhan:~/dev/psycopg3$ python tests/scripts/bench-411.py -c1 -n 1000 --no-drop --no-create psycopg2 psycopg asyncpg
2023-01-01 16:21:20,697 INFO Running psycopg2
2023-01-01 16:21:20,697 INFO thread 0 running 1000 queries
2023-01-01 16:21:23,521 INFO Run psycopg2 in 2.8213640926405787 s
2023-01-01 16:21:23,576 INFO Running psycopg sync
2023-01-01 16:21:23,576 INFO thread 0 running 1000 queries
2023-01-01 16:21:26,423 INFO Run psycopg in 2.8441125131212175 s
2023-01-01 16:21:26,442 INFO Running asyncpg
2023-01-01 16:21:26,442 INFO task 0 running 1000 queries
2023-01-01 16:21:29,422 INFO Run asyncpg in 2.9718069997616112 s

(.venv) piro@sherekhan:~/dev/psycopg3$ python tests/scripts/bench-411.py -c4 -n 1000 --no-drop --no-create psycopg2 psycopg asyncpg
2023-01-01 16:21:39,648 INFO Running psycopg2
2023-01-01 16:21:39,648 INFO thread 0 running 1000 queries
2023-01-01 16:21:39,648 INFO thread 1 running 1000 queries
2023-01-01 16:21:39,648 INFO thread 2 running 1000 queries
2023-01-01 16:21:39,649 INFO thread 3 running 1000 queries
2023-01-01 16:21:42,252 INFO Run psycopg2 in 2.6015874161385 s
2023-01-01 16:21:42,271 INFO Run psycopg2 in 2.620784340891987 s
2023-01-01 16:21:42,271 INFO Run psycopg2 in 2.6203282712958753 s
2023-01-01 16:21:42,275 INFO Run psycopg2 in 2.6239470690488815 s
2023-01-01 16:21:42,328 INFO Running psycopg sync
2023-01-01 16:21:42,328 INFO thread 0 running 1000 queries
2023-01-01 16:21:42,329 INFO thread 1 running 1000 queries
2023-01-01 16:21:42,329 INFO thread 2 running 1000 queries
2023-01-01 16:21:42,330 INFO thread 3 running 1000 queries
2023-01-01 16:21:44,905 INFO Run psycopg in 2.5738684562966228 s
2023-01-01 16:21:44,912 INFO Run psycopg in 2.5802711420692503 s
2023-01-01 16:21:44,931 INFO Run psycopg in 2.600018841214478 s
2023-01-01 16:21:44,954 INFO Run psycopg in 2.6233259476721287 s
2023-01-01 16:21:44,981 INFO Running asyncpg
2023-01-01 16:21:44,981 INFO task 0 running 1000 queries
2023-01-01 16:21:44,982 INFO task 1 running 1000 queries
2023-01-01 16:21:44,983 INFO task 2 running 1000 queries
2023-01-01 16:21:44,983 INFO task 3 running 1000 queries
2023-01-01 16:21:47,670 INFO Run asyncpg in 2.677923149894923 s
2023-01-01 16:21:47,685 INFO Run asyncpg in 2.69172420585528 s
2023-01-01 16:21:47,693 INFO Run asyncpg in 2.7013362641446292 s
2023-01-01 16:21:47,696 INFO Run asyncpg in 2.7037139711901546 s
```

Just running the test between two computers in the same data center (without introducing the pg_sleep delay), with a sub-ms latency, shows that latency dominates:

```
# ping -c3 SERVER
PING SERVER (IP) 56(84) bytes of data.
64 bytes from SERVER (IP): icmp_seq=1 ttl=63 time=0.389 ms
64 bytes from SERVER (IP): icmp_seq=2 ttl=63 time=0.365 ms
64 bytes from SERVER (IP): icmp_seq=3 ttl=63 time=0.418 ms

--- SERVER ping statistics ---
3 packets transmitted, 3 received, 0% packet loss, time 2018ms
rtt min/avg/max/mdev = 0.365/0.390/0.418/0.021 ms

# python tests/scripts/bench-411.py -c1 -n 10000 --no-drop --no-create psycopg psycopg2 asyncpg
2023-01-01 16:47:39,998 INFO Running psycopg sync
2023-01-01 16:47:39,998 INFO thread 0 running 10000 queries
2023-01-01 16:47:43,913 INFO Run psycopg in 3.902119679376483 s
2023-01-01 16:47:43,926 INFO Running psycopg2
2023-01-01 16:47:43,927 INFO thread 0 running 10000 queries
2023-01-01 16:47:47,843 INFO Run psycopg2 in 3.8995087668299675 s
2023-01-01 16:47:47,881 INFO Running asyncpg
2023-01-01 16:47:47,882 INFO task 0 running 10000 queries
2023-01-01 16:47:51,835 INFO Run asyncpg in 3.936627925373614 s

# python tests/scripts/bench-411.py -c4 -n 10000 --no-drop --no-create psycopg psycopg2 asyncpg
2023-01-01 16:48:21,847 INFO Running psycopg sync
2023-01-01 16:48:21,848 INFO thread 0 running 10000 queries
2023-01-01 16:48:21,852 INFO thread 1 running 10000 queries
2023-01-01 16:48:21,856 INFO thread 2 running 10000 queries
2023-01-01 16:48:21,860 INFO thread 3 running 10000 queries
2023-01-01 16:48:26,959 INFO Run psycopg in 5.074241460300982 s
2023-01-01 16:48:27,011 INFO Run psycopg in 5.136246461421251 s
2023-01-01 16:48:27,047 INFO Run psycopg in 5.171854223124683 s
2023-01-01 16:48:27,076 INFO Run psycopg in 5.186801585368812 s
2023-01-01 16:48:27,088 INFO Running psycopg2
2023-01-01 16:48:27,089 INFO thread 0 running 10000 queries
2023-01-01 16:48:27,092 INFO thread 1 running 10000 queries
2023-01-01 16:48:27,095 INFO thread 2 running 10000 queries
2023-01-01 16:48:27,099 INFO thread 3 running 10000 queries
2023-01-01 16:48:34,162 INFO Run psycopg2 in 7.0010702619329095 s
2023-01-01 16:48:34,178 INFO Run psycopg2 in 7.02390216384083 s
2023-01-01 16:48:34,185 INFO Run psycopg2 in 7.021626316010952 s
2023-01-01 16:48:34,190 INFO Run psycopg2 in 7.036228279583156 s
2023-01-01 16:48:34,219 INFO Running asyncpg
2023-01-01 16:48:34,219 INFO task 0 running 10000 queries
2023-01-01 16:48:34,223 INFO task 1 running 10000 queries
2023-01-01 16:48:34,225 INFO task 2 running 10000 queries
2023-01-01 16:48:34,228 INFO task 3 running 10000 queries
2023-01-01 16:48:39,225 INFO Run asyncpg in 4.9700628509745 s
2023-01-01 16:48:39,260 INFO Run asyncpg in 5.005125615745783 s
2023-01-01 16:48:39,291 INFO Run asyncpg in 5.035389132797718 s
2023-01-01 16:48:39,317 INFO Run asyncpg in 5.063969322480261 s
```

Actually, psycopg 3 seems to add less overhead in parallel queries, compared to psycopg2, and it's on par with asyncpg.

All this doesn't even mention batch mode. In a quick test in the same data center, testing batch mode (ignoring the results) looks like:

```
(.venv) root@45-33-36-154:~/psycopg# python tests/scripts/bench-411.py -c1 -n 10000 --no-drop --no-create psycopg psycopg_batch psycopg2 asyncpg
2023-01-01 16:57:28,139 INFO Running psycopg sync
2023-01-01 16:57:28,140 INFO thread 0 running 10000 queries
2023-01-01 16:57:31,791 INFO Run psycopg in 3.5337226213887334 s
2023-01-01 16:57:31,793 INFO Running psycopg batch
2023-01-01 16:57:31,793 INFO thread 0 running 10000 queries
2023-01-01 16:57:32,597 INFO Run psycopg_batch in 0.7916216999292374 s
2023-01-01 16:57:32,605 INFO Running psycopg2
2023-01-01 16:57:32,606 INFO thread 0 running 10000 queries
2023-01-01 16:57:36,450 INFO Run psycopg2 in 3.8332644449546933 s
2023-01-01 16:57:36,476 INFO Running asyncpg
2023-01-01 16:57:36,476 INFO task 0 running 10000 queries
2023-01-01 16:57:40,257 INFO Run asyncpg in 3.7656691232696176 s
```

<details>
<summary>(benchmark modified with this patch)</summary>

```patch
diff --git a/tests/scripts/bench-411.py b/tests/scripts/bench-411.py
index a4d1af30..15daf0b4 100644
--- a/tests/scripts/bench-411.py
+++ b/tests/scripts/bench-411.py
@@ -21,6 +21,7 @@ class Driver(str, Enum):
     psycopg2 = "psycopg2"
     psycopg2_green = "psycopg2_green"
     psycopg = "psycopg"
+    psycopg_batch = "psycopg_batch"
     psycopg_async = "psycopg_async"
     asyncpg = "asyncpg"
 
@@ -71,6 +72,11 @@ def main() -> None:
 
             run_psycopg(psycopg, args)
 
+        elif name == Driver.psycopg_batch:
+            import psycopg
+
+            run_psycopg_batch(psycopg, args)
+
         elif name == Driver.psycopg_async:
             import psycopg
 
@@ -150,8 +156,6 @@ def run_psycopg2(psycopg2: Any, args: Namespace) -> None:
                 for id_ in to_query:
                     with conn.cursor() as cursor:
                         cursor.execute(select, {"id": id_})
-                        cursor.fetchall()
-                    # conn.rollback()
 
     if args.concurrency == 1:
         run(0)
@@ -228,8 +232,6 @@ def run_psycopg(psycopg: Any, args: Namespace) -> None:
                 for id_ in to_query:
                     with conn.cursor() as cursor:
                         cursor.execute(select, {"id": id_})
-                        cursor.fetchall()
-                    # conn.rollback()
 
     if args.concurrency == 1:
         run(0)
@@ -245,6 +247,26 @@ def run_psycopg(psycopg: Any, args: Namespace) -> None:
             conn.commit()
 
 
+def run_psycopg_batch(psycopg: Any, args: Namespace) -> None:
+    logger.info("Running psycopg batch")
+
+    def run(i):
+        logger.info(f"thread {i} running {args.ntests} queries")
+        to_query = random.choices(ids, k=args.ntests)
+        with psycopg.connect(args.dsn) as conn:
+            with time_log("psycopg_batch"):
+                with conn.pipeline():
+                    for id_ in to_query:
+                        with conn.cursor() as cursor:
+                            cursor.execute(select, {"id": id_})
+
+    if args.concurrency == 1:
+        run(0)
+    else:
+        with ThreadPoolExecutor(max_workers=args.concurrency) as executor:
+            list(executor.map(run, range(args.concurrency)))
+
+
 async def run_psycopg_async(psycopg: Any, args: Namespace) -> None:
     logger.info("Running psycopg async")
 
@@ -308,10 +330,7 @@ async def run_asyncpg(asyncpg: Any, args: Namespace) -> None:
         conn = await asyncpg.connect(args.dsn)
         with time_log("asyncpg"):
             for id_ in to_query:
-                # tr = conn.transaction()
-                # await tr.start()
-                await conn.fetch(a_select, id_)
-                # await tr.rollback()
+                await conn.execute(a_select, id_)
         await conn.close()
 
     if args.concurrency == 1:
```

</details>

I would like to summarise these finding properly, because there is a lot of misconception about benchmarking a database driver. Asyncpg homepage still boasts their adapter to be 3x faster than psycopg2... I find it hard to believe, unless it's a query on localhost.

**Comment by pwtail** [comment]
@dvarrazzo Thank you very much. Can't wait to read the summary

**Comment by dlax** [comment]
@dvarrazzo, I can confirm you last results.

On my side, I changed the local latency using [tc](https://manpages.debian.org/bullseye/iproute2/tc.8.en.html), instead of `pg_sleep()`:
```
$ sudo tc qdisc add dev lo root netem delay 300
$ ping localhost
PING localhost(localhost (::1)) 56 data bytes
64 bytes from localhost (::1): icmp_seq=1 ttl=64 time=0.632 ms
64 bytes from localhost (::1): icmp_seq=2 ttl=64 time=0.830 ms
64 bytes from localhost (::1): icmp_seq=3 ttl=64 time=0.817 ms
64 bytes from localhost (::1): icmp_seq=4 ttl=64 time=0.854 ms
^C
```
Then:
```
$ python tests/scripts/bench-411.py -c4 -n 10000 --no-drop --no-create psycopg2 psycopg asyncpg
2023-01-01 22:11:16,264 INFO Running psycopg2
2023-01-01 22:11:16,265 INFO thread 0 running 10000 queries
2023-01-01 22:11:16,266 INFO thread 1 running 10000 queries
2023-01-01 22:11:16,269 INFO thread 2 running 10000 queries
2023-01-01 22:11:16,270 INFO thread 3 running 10000 queries
2023-01-01 22:11:25,348 INFO Run psycopg2 in 9.061737809000988 s
2023-01-01 22:11:25,351 INFO Run psycopg2 in 9.064794611000252 s
2023-01-01 22:11:25,387 INFO Run psycopg2 in 9.098787585999162 s
2023-01-01 22:11:25,387 INFO Run psycopg2 in 9.097975227999996 s
2023-01-01 22:11:25,453 INFO Running psycopg sync
2023-01-01 22:11:25,454 INFO thread 0 running 10000 queries
2023-01-01 22:11:25,455 INFO thread 1 running 10000 queries
2023-01-01 22:11:25,456 INFO thread 2 running 10000 queries
2023-01-01 22:11:25,457 INFO thread 3 running 10000 queries
2023-01-01 22:11:33,283 INFO Run psycopg in 7.803626929000529 s
2023-01-01 22:11:33,288 INFO Run psycopg in 7.808878117999484 s
2023-01-01 22:11:33,290 INFO Run psycopg in 7.812252437001007 s
2023-01-01 22:11:33,290 INFO Run psycopg in 7.810482503000458 s
2023-01-01 22:11:33,309 INFO Running asyncpg
2023-01-01 22:11:33,309 INFO task 0 running 10000 queries
2023-01-01 22:11:33,310 INFO task 1 running 10000 queries
2023-01-01 22:11:33,312 INFO task 2 running 10000 queries
2023-01-01 22:11:33,315 INFO task 3 running 10000 queries
2023-01-01 22:11:41,779 INFO Run asyncpg in 8.33641967999938 s
2023-01-01 22:11:41,783 INFO Run asyncpg in 8.341571076000037 s
2023-01-01 22:11:41,787 INFO Run asyncpg in 8.345150424000167 s
2023-01-01 22:11:41,790 INFO Run asyncpg in 8.347713839999415 s
```
to be compared with no latency (reset with `sudo tc qdisc del dev lo root netem`):
```
$ ping localhost
PING localhost(localhost (::1)) 56 data bytes
64 bytes from localhost (::1): icmp_seq=1 ttl=64 time=0.028 ms
64 bytes from localhost (::1): icmp_seq=2 ttl=64 time=0.031 ms
64 bytes from localhost (::1): icmp_seq=3 ttl=64 time=0.031 ms
^C
$ python tests/scripts/bench-411.py -c4 -n 10000 --no-drop --no-create psycopg2 psycopg asyncpg
2023-01-01 22:16:48,585 INFO Running psycopg2
2023-01-01 22:16:48,585 INFO thread 0 running 10000 queries
2023-01-01 22:16:48,586 INFO thread 1 running 10000 queries
2023-01-01 22:16:48,588 INFO thread 2 running 10000 queries
2023-01-01 22:16:48,591 INFO thread 3 running 10000 queries
2023-01-01 22:16:50,548 INFO Run psycopg2 in 1.9453500640011043 s
2023-01-01 22:16:50,552 INFO Run psycopg2 in 1.9456397670001024 s
2023-01-01 22:16:50,566 INFO Run psycopg2 in 1.9583636740007933 s
2023-01-01 22:16:50,576 INFO Run psycopg2 in 1.9746641340007045 s
2023-01-01 22:16:50,642 INFO Running psycopg sync
2023-01-01 22:16:50,643 INFO thread 0 running 10000 queries
2023-01-01 22:16:50,647 INFO thread 1 running 10000 queries
2023-01-01 22:16:50,647 INFO thread 2 running 10000 queries
2023-01-01 22:16:50,650 INFO thread 3 running 10000 queries
2023-01-01 22:16:55,723 INFO Run psycopg in 5.052449745000558 s
2023-01-01 22:16:55,741 INFO Run psycopg in 5.071528365000631 s
2023-01-01 22:16:55,745 INFO Run psycopg in 5.074883404000502 s
2023-01-01 22:16:55,747 INFO Run psycopg in 5.07808085599936 s
2023-01-01 22:16:55,766 INFO Running asyncpg
2023-01-01 22:16:55,766 INFO task 0 running 10000 queries
2023-01-01 22:16:55,767 INFO task 1 running 10000 queries
2023-01-01 22:16:55,769 INFO task 2 running 10000 queries
2023-01-01 22:16:55,773 INFO task 3 running 10000 queries
2023-01-01 22:16:59,204 INFO Run asyncpg in 3.3378875569997035 s
2023-01-01 22:16:59,206 INFO Run asyncpg in 3.3116932769989944 s
2023-01-01 22:16:59,207 INFO Run asyncpg in 3.3135444350009493 s
2023-01-01 22:16:59,209 INFO Run asyncpg in 3.342617025000436 s
```

So, yes, it seems that psycopg2 performs better artificially *when* there is no latency but as long as there is a bit of latency, there is no clear winner.

> All this doesn't even mention batch mode. In a quick test in the same data center, testing batch mode (ignoring the results) looks like

ahah, not so surprising!

**Comment by dlax** [comment]
@dvarrazzo considering latest insights about benchmarks vs latency, is the plan proposed [earlier](https://github.com/psycopg/psycopg/issues/448#issuecomment-1366698491) still valid *as is*?
(I'm trying to review things but I am having hard time following things sometimes... Still, I don't mind spending more time on this *if this worth*.)

**Comment by dlax** [comment]
I had a first commit-by-commit look at #469:

* The separation of `...gen_no_pipeline()` and `...gen_pipeline()` method is a good thing. In fact, I had been thinking about a refactoring for the same goal a few months ago, but based on an intermediate object, and `ExecuteX` and a `PipelineX` variant (subclass): this object would hold all generator methods, either the non-pipeline ones or the pipeline ones, it would be attached to the connection and replaced upon enter/exit of the pipeline, thus removing all (or most) the `if self.(_conn.)_pipeline:`. That's an idea I could pursue, if you think it's worth?

* I like the renaming of `generators.execute()` as `flush_and_fetch()`.

* I'm not sure I quite get the changes about GIL release optimization; but I probably need to have a closer look.

* Some changes are (or seem) unrelated to the "high level generators" thing, such as 0458c9e5e42d648e62ef56768e78f0d86addcfd6, 757fc336d51940a0ee02461859278d258f3ab4ed, 0edfae2cb6af6fde43725f36408893a4b712fb2e and could be made into one or more separate PRs maybe? (These commits rebase fine on master, fwiw.)

* I understand this stems from an experiment, which explains that consistency between the Python and C implementation is uncoupled in the middle of the commit stack. How you would like to proceed in this respect: fold all changes? Rebase/reorganize commits?

I think we need to find a workflow for communicating around these changes; I am suggesting smaller PRs for parts of the overall plan that are ready, it'd certainly be easier to read for me and I can be more helpful then. Suggesting this because that's how I'm used to work, but if you have something else in mind, I'm open.

**Comment by dvarrazzo** [comment]
@dlax 

> I think we need to find a workflow for communicating around these changes; I am suggesting smaller PRs for parts of the overall plan that are ready, it'd certainly be easier to read for me and I can be more helpful then.

Yes, I agree in principle, and I am trying to agree in deeds too. I am trying to disentangle this MR into commits that make sense to review on their own.

**Comment by ofek** [comment]
Has there been any update on this?

**Comment by dvarrazzo** [comment]
@ofek as per description, there is no performance degradation unless running microbenchmarks on localhost. Leaving this task open because there were some ideas that can be implemented in 3.2, but the trunk is undergoing some heavy refactoring now, so changes will have to be postponed.

There are minimal improvement that can be made, but they are totally dwarfed by network time.

**Comment by ofek** [comment]
Okay, thank you for the clarification! Is there a tracking issue for the ideas proposed here?

**Comment by dvarrazzo** [comment]
Closing this branch because there is no real degradation under concurrency unless we are talking about fast queries running on localhost.

Approaches to reduce concurrency pressure involve major rewrites for marginal gains. Many of the attempts are described in [this comment](https://github.com/psycopg/psycopg/issues/448#issuecomment-1366698491). They may be revived in the future, after reviewing them for the changes that in the meantime happened to the codebase since they were opened, but this is not to consier an issue imo.

**Comment by dvarrazzo** [comment]
Related branch: https://github.com/dvarrazzo/psycopg/tree/fix-448 with more parallelism benchmarks.
